docs/io/unified.rst

```diff
@@ -642,24 +637,60 @@ overwriting existing files. To overwrite only a single table within an HDF5
+used to ensure that the data is compressed on disk::
 
-    >>> t.write('observations.hdf5', path='updated_data', serialize_meta=True)
+    >>> t.write('new_file.hdf5', path='updated_data', compression=True)
 
-The way serialized meta are saved in the HDF5 dataset have changed in Astropy 3.0.
-Files in the old format are still read correctly. If for some reason the user wants to *write*
-in the old format, they will specify the (deprecated) ``compatibility_mode`` keyword
+Metadata and mixin columns
+^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-    >>> t.write('observations.hdf5', path='updated_data', serialize_meta=True, compatibility_mode=True)
+Astropy tables can contain metadata, both in the table ``meta`` attribute
+(which is an ordered dictionary of arbitrary key/value pairs), and within the
+columns, which each have attributes ``unit``, ``format``, ``description``,
+and ``meta``.
 
-Finally, when writing to HDF5 files, the ``compression=`` argument can be
-used to ensure that the data is compressed on disk::
+By default, when writing a table to HDF5 the code will attempt to store each
+key/value pair within the table ``meta`` as HDF5 attributes of the table
+dataset.  This will fail of the values within ``meta`` are not objects that can
+be stored as HDF5 attributes.  In addition, if the table columns being stored
+have defined values for any of the above-listed column attributes, these
+metadata will *not* be stored and a warning will be issued.
 
-    >>> t.write('new_file.hdf5', path='updated_data', compression=True)
+serialize_meta
+""""""""""""""
+To enable storing all table and column metadata to the HDF5 file, call
+the ``write()`` method with ``serialize_meta=True``.  This will store metadata
+in a separate HDF5 dataset, contained in the same file, which is named
+``<path>.__table_column_meta__``.  Here ``path`` is the argument provided in
+the call to ``write()``::
+
+    >>> t.write('observations.hdf5', path='data', serialize_meta=True)
+
+As of astropy 3.0, by specifying ``serialize_meta=True`` one can also store to
+HDF5 tables that contain :ref:`mixin_columns` such as `~astropy.time.Time` or
+`~astropy.coordinates.SkyCoord` columns.
+
+compatibility_mode
+""""""""""""""""""
+
+The way metadata are saved in the HDF5 dataset has changed in astropy 3.0.
+Previously the metadata were serialized with YAML and this was stored as an
+HDF5 attribute.  This process was subject to a fixed limit on the size of an
+attribute.  Starting with 3.0 the YAML-serialized metadata are stored as a
+separate dataset as described above, with no size limit.
+
+Files using the old convention are automatically recognized and will always be read
+correctly.
+
+If for some reason the user needs to *write* in the old format, they should
+specify the deprecated ``compatibility_mode`` keyword::
 
+    >>> t.write('observations.hdf5', path='updated_data', serialize_meta=True,
+    ...         compatibility_mode=True)
```

Comment:
Not directly related to this PR, but might it be an idea to use a string here instead of a bool, so that it is a bit more future-proof? (eg., `compatibility_mode='2.0'` -- we recently did something similar in `np.array2print`, with a `legacy` argument that takes a string).

...

Although given the below warning, this is probably useless.

Reply:
I thought about this a bit and do generally like the idea of a version number instead of bool.  But in the end I would be quite surprised if there is anybody writing HDF5 files with astropy and using the contents of that meta attribute for anything but round-tripping back into astropy.  And if they are doing so, they already have the machinery for decoding YAML and handling it.  The change is pretty straightforward (as shown in our reader code) because the content of the YAML is the same, it's only where it gets stored that changed.

Upshot - we should just take this out after a release or two so we can clean up the code and API.  It's already not so pretty and I wish we didn't have to use `serialize_meta=True` every time now for mixin support.  This is confusing, and coupled with the converse problem in FITS of requiring `astropy_native=True` always...  Maintaining back-compatibility can drive the API in annoying directions.

Which is another question, namely changing the default for `serialize_meta` to `True`.  I think this release is too early but we might go there in the future.

Reply:
OK, makes sense. Perhaps we can aim for changing it in the next LTS.